server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -791,6 +780,9 @@ private void executePipelines(
                             slot,
                             new IllegalStateException("final pipeline [" + pipelineId + "] can't change the target index")
                         );
+                        // document failed! no further processing of this doc
+                        onFinished.onResponse(null);
```

Comment:
Seems strange that we have this pattern of calling onFailure.accept() and then you have to call onFinished.onResponse(). And we never call onFinished.onFailure(). What if we wrap these two things in an ActionListener, so there's just one thing to pass around, and you can't screw up and call the wrong sequence of listeners/consumers? Or do you think that's one level of listeners too far? The big downside is that you have to have one per slot (since the slot number is passed to the failure handler).

Reply:
Yeah! It is weird. I’m on board with unifying the three listeners/handlers into one, and while this is surely a small step in that direction I’d like to keep that off this PR for now.

Okay if you and I look into that next step in January?

Reply:
Sounds good. This is definitely an improvement over what we have.

Reply:
It's me from the future, this ended up being a part of https://github.com/elastic/elasticsearch/pull/93329. 😄